ADR 028: Public Key Addresses

aaronc opened the pull request:
ref: #5694 

This ADR proposes a canonical extensible address format for new public key types, and addresses the multisig pubkey address issue.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

alexanderbez commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+
+In discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we agreed to go with an
+approach similar to this where essentially we take the first 20 bytes of the `sha256` hash of
+the key type concatenated with the key bytes, summarized as `Sha256(KeyTypePrefix || Length || Keybytes)[:20]`.
```
Note, you include `length` here, but not in the `Decision` below.

aaronc replied:
I quoted directly from the issue, but I'll take it out.

alexanderbez commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+Following on the discussion in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we propose the
+following approach.
+
+All public key types will have a unique protobuf message type such as:
```
If I understand this correctly, this states that new (non-secp256k1) keys will take this format and existing secp256k1 keys will adhere to the current design?

ethanfrey commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+
+## Decision
+
+### Legacy Public Key Addresses Don't Change
```
This is essential. It adds complexity but needed for migration. Maybe we can import only a minimal subset of amino needed for this.

ethanfrey commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+
+For all basic public keys, key bytes should just be the raw unencoded public key bytes.
+
+### Multisig Addresses
```
I like this one (as well as those above).
One canonical form, well documented

ethanfrey reviewed (commented):
Sounds good

clevinson commented:
The proposal here looks good to me, in terms of the specification for new addresses.  However, I'm a bit confused how this actually relates to existing address formats, and I think the wording around that could be clarified.

Just so I understand correctly, this means that legacy account addresses would be unchanged and remain in their bech32 prefixed form, but accounts generated with non `secp256k1` pubkeys would use the new format?

Similarly, what constitutes a "legacy" or "new" multisig? Is this about addresses that exist onchain, vs ones that do not? Or about what pubkey type the multisig address is comprised of?

amaurymartiny commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+### Legacy Public Key Addresses Don't Change
+
+`secp256k1` and multisig public keys are currently in use in existing Cosmos SDK zones. We
+don't want to change existing addresses. So the addresses for these two key types will remain the same.
```
as cory pointed out, I think it would make this ADR clearer to state what these 2 key types use now:
- secp256k1: `ripemd160(sha256(pk_bytes))`
- legacy amino multisig: `sha256(aminoCdc.Marshal(pk))[:20]`

amaurymartiny commented on docs/architecture/adr-028-public-key-addresses.md:
````diff
@@ -0,0 +1,110 @@
+}
+```
+ 
+All protobuf messages have unique fully qualified names, in this example `cosmos.crypto.sr25519.PubKey`.
````
If we strictly follow buf's linting rules for package names (version-suffixed), we might have `cosmos.crypto.sr25519.v1.PubKey`.

I would advocate to break those linting rules here, `cosmos.crypto.sr25519.PubKey` (without v1) reads much better.

robert-zaremba replied:
Is this going to be an algorithm to construct addresses for any asymetric key schema?

amaurymartiny replied:
Yes, see #5694 for motivation.

amaurymartiny commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+key type prefix when creating addresses.
+
+We define the canonical address format for new (non-legacy) public keys as
+`Sha256(fmt.Sprintf("%s/%x, proto.MessageName(key), key.Bytes())[:20]`. This takes
```
Current secp256k1 uses ripemd160 instead of [:20]. The latter is faster to compute, but are there other trade-offs (in particular, security-wise)?

ethanfrey replied:
It is needed for bitcoin compatibility. There were other designs, but the 2017 fundraiser addresses were generated with a bitcoin secp256k1 address derivation and the go code had to be compatible.

But I cannot tell you if 20 bytes prefix is just as safe

zmanian replied:
ripemd160 is an odd and basically pointless cryptographic choice and there is no reason to retain it.

It's a slow hash function and doesn't improve security in a meaningful way.

aaronc replied:
Why not blake3 @robert-zaremba ?

robert-zaremba replied:
I'm all for `blake3` (especially that it's officially endorsed on the blake2 side) but in other discussion majority was afraid that it's too fresh. https://github.com/cosmos/cosmos-sdk/pull/7259#issuecomment-688896136

aaronc replied:
I'd be in favor of blake3 if we're going that route. I think the issue regarding state sync was more that the cryptography is a relatively minor part of snapshot sync. Whereas with address formats it's more critical. So I think that would merit it taking a more central role and us attempting to choose the best algorithm, which by everything I can see appears to be blake3.

aaronc replied:
Although, I would say that the downside that needs to be considered is the availability of a high-quality implementation. Clients should be able to compute addresses from public keys and sha256 pretty ubiquitous. Address hashing is also done over a very small piece of data so I expect the performance benefits would be minimal. It probably only makes sense if there is some security improvement to a different algorithm. I would expect we'll see more of a performance effect if we're looking at block or merkle tree hashing.

robert-zaremba replied:
`blake2b` is considered safer than `sha2`. Changing address creation algorithm is way more complex thing than chunks hashes.

aaronc replied:
> `blake2b` is considered safer than `sha2`.

Can you articulate the safety benefits `blake2b` or `blake3` would offer for address hasing?

> Changing address creation algorithm is way more complex thing than chunks hashes.

Yes, I totally agree. What we define here will (hopefully) be in use for a long time.

robert-zaremba replied:
> BLAKE2b is faster than MD5, SHA-1, SHA-2, and SHA-3, on 64-bit x86-64 and ARM architectures.[3] BLAKE2 provides security superior to SHA-2 and similar to that of SHA-3: immunity to length extension, indifferentiability from a random oracle, etc.

source: [wikipedia](https://en.wikipedia.org/wiki/BLAKE_%28hash_function%29#BLAKE2)

Blake2 is conjectured to be random oracle indifferentiable, a stronger property that SHA-256 doesn't have, and which implies the three classic hash function properties.

AFAIK, blake3 adds parallelism to the design goals, so it's much faster.

Computing an address is something we will do often. In general, my approach is: if something is validated to work correctly and is better than we should just use it.

aaronc replied:
So hypothetically it would be easier to find a pre-image for a SHA256 address than a Blake2 address?

I would note that Blake3 reduced the number of rounds from Blake2 to get a speed up. We should review the academic assessment of that decision because it appears a bit controversial. For something like this where performance is less critical, maybe we err on the side of caution and go with Blake2?

robert-zaremba replied:
> So hypothetically it would be easier to find a pre-image for a SHA256 address than a Blake2 address?

Not worse, and it's more random.

robert-zaremba replied:
> Blake3 reduced the number of rounds from Blake2 to get a speed up

It uses a different mechanism, so I can't say if the reduction of number of rounds impacts the security vs Blake2 or not. Probably asking one of the authors (Zooko?) would be the best approach.

robert-zaremba replied:
> maybe we err on the side of caution and go with Blake2

This seams to be the balanced solution. Although it would be still good to check out with zcash about their view. I will try to reach out to some of them.

aaronc replied:
The new draft uses `blake2b`. Both because it has better library support and has been around longer and vetted more thoroughly (as I understand it).

I'm assuming `blake2b` is preferable to `blake2s` for our use case?

alain2sf replied:
BLAKE2b is designed and optimized for 64-bit CPUs. BLAKE2s is designed and optimized for 8, 16, 32-bit CPUs. 
BLAKE2 has indeed excellent Golang and Rust libraries support. Today there are very few BLAKE3 supported implementations. The speed boost observed with BLAKE3 is not really a selling point regarding ADR 028.

amaurymartiny reviewed (commented):
From the comments in this ADR and in the linked issues, the feeling I get is that everyone here is aligned to say "it should be okay" (incl. me).

Would it make sense to run this by a cryptographer, to get validation "yes, this is okay". I'd trust the word of a cryptographer over my gut feeling any day of the week.

amaurymartiny commented:
@clevinson I can try to answer your questions, given my understanding of this ADR:

> If i understand correctly, this means that legacy account addresses would be unchanged and remain in their bech32 prefixed form, but accounts generated with non secp256k1 pubkeys would use the new format (no prefix, just a hash as described above)?

There's public key (`[]byte`), address (also `[]byte`, derived from pubkey & keytype, what we're discussing here) and bech32-encoding of an address (e.g. a `cosm`-prefixed string). Whatever address format is chosen here, we can still bech32-encode it.

> How does one distinguish between a "legacy" or "new" multisig? Is this about addresses that exist onchain on a current zone, vs ones that do not? Or about what pubkey type the multisig address is comprised of (where secp256k1 is deemed "legacy"). How does that then relate to multisig with keys of multiple types?

Given a random 20-byte address, I believe there's no way to know if it's a "legacy" or "new" multisig, or any other single public key type. This ADR describes the other direction, given 1/ threshold and 2/ an array of public keys (each being potentially of a different key type), how to construct "legacy" and "new" multisig addresses.

Also, I don't think secp256k1 is deemed "legacy"; it just has a different address generation algorithm.

robert-zaremba commented:
For new account `address` format, how about using the same algorithm as Algorand is [using](https://developer.algorand.org/docs/features/accounts/#transformation-public-key-to-algorand-address).

They have one of the best cryptographers form the world so they know best what / how to use.

alessio commented:
The proposal looks good, though I'm interested in understanding better what are the consequences for multisig keys (see Cory's questioin: `How does one distinguish between a "legacy" or "new" multisig? `)

robert-zaremba commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+## Status
+
+Proposed
+
```
Let's add a `Summary` or `Abstract` section right after Changelog and Status.

aaronc replied:
Could you make a proposal around this to the ADR template itself? I'm just trying to follow what's in the template.

clevinson commented:
This should get a minor update once #7109 / #6886 are completed. But probably won't be implemented until 0.41 @aaronc

ebuchman commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+
+### Neutral
+- protobuf message names are used as key type prefixes
+- public key bytes are hex encoded before generating addresses
```
A bit awkward and maybe should be better motivated if its necessary.

aaronc replied:
Yep, removed hex encoding.

ebuchman commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+We define the canonical address format for new (non-legacy) public keys as
+`Sha256(fmt.Sprintf("%s/%x, proto.MessageName(key), key.Bytes())[:20]`. This takes
+the first 20 bytes of an SHA-256 hash of a string with the proto message name for the key
+type joined by an `/` with the hex encoding of the key bytes.
```
Actually, why bother with the `/` either? In #5694 we talked about `Sha256(KeyTypePrefix || Keybytes)[:20]` which is just pure concatenation. I guess because the prefix is a string not just a byte, so we need a domain separator? It seems like maybe we're optimizing for human readability of the pre-image; is that important for something?

robert-zaremba replied:
PubKeys are not meant to be human readable. They are long. That's why we are creating addresses.

So we don't even need to convert to string. We can concatenate bytes, without conversion to string.

aaronc replied:
I've updated this ADR to define an `AddressHash` function which takes a string `prefix` and binary `contents` separating them by a null terminator `0` byte and which doesn't do hex-encoding of binary. Can you take a look again @ebuchman ?

ebuchman reviewed (commented):
Nice! Just some questions around the hex encoding in the pre-image

ebuchman commented:
> For new account address format, how about using the same algorithm as Algorand is using.

> Additional advantage is that it has a checksum, which protects against a wrong copy paste!

Looks like they're just using `PubKey || checksum(PubKey)` as the address and base32 encoding it.

That's a long address. We don't want to break our existing standard for 20-byte addresses. We also only show users bech32 forms of the address (rather than the raw 20 bytes), which include a checksum. True, the checksum isn't part of the protocol, its just part of the UX.

ebuchman commented:
Another question here - are there possible conflicts between Amino-legacy and new multisig formats? Could the same set of publickeys+threshold now effectively get a second address? Are there any risks in that?

robert-zaremba reviewed (changes requested):
Let's make sure (or at least lets leave a note in the document to iterate on it in the future) that we understand the security problem here.

robert-zaremba commented:
The thing we want to achieve here is to find a secure algorithm to create short addresses.

Let's add some requirements: 

1. Do we want to encode in an address a PK scheme (so having a function: `addressType: Address -> AddressType`)? I think we don't need to, but it may be required for security reason.
2. In #3685  we wanted to have a variable address length. It turns out that we don't want it any more?
3. We want not malleable addresses. 
    * With the algorithm, which is proposed here, I think that adding a prefix before hashing doesn't change anything: probability of finding 2 colliding addresses in the same PK scheme is the same as the probability of finding 2 colliding addresses in 2 different PK scheme (as explained in the second point of https://github.com/cosmos/cosmos-sdk/pull/7086#discussion_r499047318)

Is there any other requirement?

robert-zaremba commented:
Also, some PK schemas will have their native Multisig PK derived from the _child_ PKs (eg BLS, Schnorr). So maybe for the multisig address we can have:

1. If a multisig scheme `M` has a native aggregated `M_PK`, then `address(M) = canonical_address(M_PK)`
2. otherwise:

        let [k1, ...., kn] = child_keys(M);
        address(M) = canonical_address(typ(M), address(pk1) || ... || address(pkn))  
        // NOTE: as in the comment above I don't think the typ(M) is necessary in canonical_address)

aaronc commented:
> 2. In #3685  we wanted to have a variable address length. It turns out that we don't want it any more?

It turns out that doing this in practice is harder than it seems. For instance if you look in `x/staking/types/keys.go` you will see that so much of this storage code depends on 20 byte addresses. This is obviously not a good security reason to stick with 20 byte addresses everywhere, but it certainly puts a significant limitation on our ability to do so without a serious refactor of the staking module and likely other places. (#7098 could help out with that significantly btw but it would still be a heavy lift.)

robert-zaremba commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,111 @@
+
+> The good example you brought up was eg. a public key bytes being a valid public key on two algorithms supported by the codec. Meaning if either was broken, you would break accounts even if they were secured with the safer variant. This is only as the issue when no differentiating type info is present in the preimage (before hashing into an address).
+
+> I would like to hear an argument if the 20 bytes space is an actual issue for security, as I would be happy to increase my address sizes in weave. I just figured cosmos and ethereum and bitcoin all use 20 bytes, it should be good enough. And the arguments above which made me feel it was secure. But I have not done a deeper analysis.
```
To be clear - I'm not saying that adding prefix is bad. I'm not sure it solves anything. @ethanfrey ?

robert-zaremba replied:
There is one more attack, which in fact is important here, and this update address it:

3. Similar to `(1.)`: `A` became vulnerable. Attacker is able to forge a signature for `pk \in A`. In all places where we don't store any relationship between addresses and `(PK, PK scheme)` pair, the attacker will be able to spend address assets.

This proposal (including `scheme` url / name in the address algorithm) protects against the attack described #3685 (isolating address spaces to protect against attacks when one of the PK scheme is broken).

robert-zaremba commented:
What about multisig  accounts with dynamic addresses? I think the solution will be to use a native multisig address if such one is provided (I proposed it in one comment).

robert-zaremba commented:
> Some PK schemas have their native Multisig PK derived from the _child_ PKs (eg BLS, Schnorr). So maybe for the multisig address we can have:
> 
>     1. If a multisig scheme `M` has a native aggregated `M_PK`, then `address(M) = canonical_address(M_PK)`
> 
>     2. otherwise:
>        ```
>         let [k1, ...., kn] = child_keys(M);
>         address(M) = canonical_address(typ(M), address(pk1) || ... || address(pkn))  
>         // NOTE: as in the comment above I don't think the typ(M) is necessary in canonical_address)
>        ```

We also need to assure a deterministic order of addresses we hash.

robert-zaremba commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+`secp256k1` and multisig public keys are currently in use in existing Cosmos SDK zones. We
+don't want to change existing addresses. So the addresses for these two key types will remain the same.
+
+The current multisig public keys use amino serialization to generate the address. We will retain
```
Or let's use a more precise language: when someone will read it in 2months he should get it right.

aaronc replied:
The next sentence specifies that they will be called "legacy amino" multisig pubkeys

robert-zaremba commented:
This ADR has status _proposed_.
Shall we wrap up this PR?

* address things we can address now (easily)
* add `TODO` in sections which we still need to work on.

And then iterate?

aaronc commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+point for forming the multisig address. Let's create an array of strings, `hexAddresses []string`,
+which is the hex-encoded address of each nested pubkey. We join these hex encoded addresses
+with a `/`, i.e. `joinedHexAddresses := strings.Join(hexAddresses, "/")`. We then form the address of the multisig pubkey,
+using `Sha256(fmt.Sprintf("cosmos.crypto.multisig.PubKey/%d/%s", pk.Threshold, joinedHexAddresses)[:20]`.
```
Because we have to deal with non-deterministic encoding of protobuf (i.e. ADR 027). This avoids that. Can you see if what I have their now which does sorting is okay?

ethanfrey commented:
I support merging this with status proposed. Then opening another PR to address minor concerns. At least the majority of the points have been clarified and it is good to make this more visible.

aaronc commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,110 @@
+- a simple algorithm for generating addresses for new public keys
+
+### Negative
+- addresses do not communicate key type, a prefixed approach would have done this
```
I don't know. You're saying that maybe we can have the bech32 have an extra prefix that isn't present in the actual address? Would this be something like `cosmossecp256k1sdgh3sghlsdsdg`. Or would the prefix get added before bech32 encoding?

aaronc commented:
> Another question here - are there possible conflicts between Amino-legacy and new multisig formats? Could the same set of publickeys+threshold now effectively get a second address? Are there any risks in that?
> 
> For instance I could imagine a replay problem if "from" address isn't signed over, but it seems that's usually included in the msgs.

I don't know. Theoretically I guess... The from address usually is signed over. Is it a problem if the same publickeys+thresholds got a second account with a different address depending on the key type?

aaronc commented:
> What about multisig accounts with dynamic addresses? I think the solution will be to use a native multisig address if such one is provided (I proposed it in one comment).

I'm not sure what you mean by this? Would group accounts (https://github.com/regen-network/cosmos-modules/tree/master/incubator/group) satisfy would you mean by dynamic multisig @robert-zaremba ?

> We also need to assure a deterministic order of addresses we hash.

I've sorted the addresses in the current multisig format. Does that address your concerns?

aaronc commented:
I've tried to address all comments and have significantly update the proposal to:
- not hex encode bytes before hashing
- use `blake2b` hashes
- address module account addresses for cases like `staking` or `group` (where modules can have any number of sub-accounts or pools)
- sort multisig keys before hashing

Please let me know what you think if you get a chance to re-review.

robert-zaremba commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -51,10 +51,43 @@ The current multisig public keys use amino serialization to generate the address
+	if len(contents) != 0 {
+		preImage = append(preImage, 0)
+		preImage = append(preImage, contents...)
+	}
```
When implementing this we should use use `make([]byte, correctSize)` and copy they content using `for` or `copy` (into slice starting from index `2`).

robert-zaremba commented:
> I've sorted the addresses in the current multisig format. Does that address your concerns?

Yes, but we will still need to rethink how to handle dynamic accounts (more about it below).

> > What about multisig accounts with dynamic addresses? I think the solution will be to use a native multisig address if such one is provided (I proposed it in one comment).
> 
> I'm not sure what you mean by this? Would group accounts (https://github.com/regen-network/cosmos-modules/tree/master/incubator/group) satisfy would you mean by dynamic multisig @robert-zaremba ?

Maybe. This will go into the category of a complex / multisig account with a native address. I think we will need to go back to the ADR28 once we will put all accounts together and before we will implement it. So for the moment I would like to mark it only.

robert-zaremba commented:
In general I would use a pseudo code for explaining the algorithm. The reason is that we should discuss implementation details separately, which will involve things like integration and optimization, and language specific things.

robert-zaremba commented on docs/architecture/adr-028-public-key-addresses.md:
```diff
@@ -0,0 +1,171 @@
+In the `preImage`, the byte value `0` is used as the separator between `prefix` and `contents`. This is a logical
+choice given that `0` is an invalid value for a string character and is commonly used as a null terminator.
+
+We use a 256-bit `blake2b` hash instead of `sha256` because it is generally considered more secure. Blake hashes
+are considered "random oracle indifferentiable", a stronger property which `sha256` does not have.
```
@zmanian  I agree with your general approach. But we don't propose a any exotic algorithm. Blake2 is part of the at least Go and Python3 stdlib and it's an improvement. Some cryptographers claim that sha256 is easier to break than blake2 because we already know how to break other hash functions from the MD family.

alessio replied:
> Some cryptographers claim that sha256 is easier to break than blake2.

Some references would be appreciated :+1:

aaronc replied:
> I'm just saying this because I'm always going to strongly advocate for a really high bar for adding a new hash function to the Cosmos trusted computing base.

So I want to summarize what I heard from @zmanian on our last call about this. Basically:
- there is no consensus on the replacement for SHA256 yet in the crypto community and there's a widespread sense that it won't be broken anytime soon
- we want to minimize what is needed for someone to implement "cosmos" on a broad array of devices including hardware wallets and embedded devices

Thus for now it's probably preferable to stick with SHA256. Is that accurate @zmanian ?

@alessio any progress on getting the cryptographer at AiB to take a look at this?

robert-zaremba replied:
W3F Research  / Polkadot is using blake2b and avoiding sha2.

aaronc commented:
I have reverted to change to `blake2b` and this ADR now uses `sha256` again. I would reconsider an alternate hash if a good argument is made, but it's contentious for now and I want to at least be able to move forward with the non-contentious parts.

In the interest of progress and the points raised in #7498 can we merge this ADR with status proposed and address remaining issues as follow-up issues and or PRs?

I still would like to get a cryptographer's opinion on the hashing if we can, but I think I have addressed what I can for now.

So basically I would like to get people's opinion as to whether this is a good enough starting point that we can move forward with it at least preliminarily. We need something to work off of within the SDK. Our plan is to include this in the v0.41 release so we can open a few follow-up issues tagged as v0.41 and make sure we address them before release.

aaronc commented:
I have reverted to change to `blake2b` and this ADR now uses `sha256` again. I would reconsider an alternate hash if a good argument is made, but it's contentious for now and I want to at least be able to move forward with the non-contentious parts.

In the interest of progress and the points raised in #7498 can we merge this ADR with status proposed and address remaining issues as follow-up issues and or PRs?

I still would like to get a cryptographer's opinion on the hashing if we can, but I think I have addressed what I can for now.

So basically I would like to get people's opinion as to whether this is a good enough starting point that we can move forward with it at least preliminarily. This has been an open issue

robert-zaremba commented:
Let's add notes around the parts which:
1. Require opinion of qualified person
2. Are still under consideration

And then merge

robert-zaremba reviewed (changes requested):
Let's add notes around the parts which:

+ Require opinion of qualified person
+ Are still under consideration

And then merge

robert-zaremba reviewed (approved):
Let's add notes around the parts which:

-     Require opinion of qualified person
-     Are still under consideration

And then merge

aaronc commented:
> Let's add notes around the parts which:
> 
> * Require opinion of qualified person
> * Are still under consideration
> 
> And then merge

@robert-zaremba could you open an issue documenting those things?
